File: Homework_3/diff_eqn_solver.f90
function f(x,y) result(z)
        double precision x, y, z
        z = -x*y
end function f

function euler_method(x0,y0,x,stepsize) result(e)
        double precision f, x0, y0, x, stepsize, x1, y1, e
        integer n, i
        n = (x - x0)/stepsize
        do 10 i = 0, n
                y1 = y0 + stepsize*f(x0, y0)
                x1 = x0 + stepsize
                x0 = x1
                y0 = y1
        10 continue
        e = y1
end function euler_method

function partialx(x0,y0,stepsize) result(ddx)
        double precision f, x0, y0, stepsize, ddx
        ddx = (f(x0-2.0*stepsize,y0)-8.0*f(x0-stepsize,y0)+8.0*f(x0 + stepsize,y0)-f(x0+2.0*stepsize,y0))/(12.0*stepsize)
end function partialx

function partialy(x0,y0,stepsize) result(ddy)
        double precision f, x0, y0, stepsize, ddy
        ddy = (f(x0,y0-2.0*stepsize)-8.0*f(x0,y0-stepsize)+8.0*f(x0,y0+stepsize)-f(x0,y0+2.0*stepsize))/(12.0*stepsize)
end function partialy

function taylor_method(x0,y0,x,stepsize) result(t)
        double precision f, x0, y0, x, stepsize, x1, y1, t, partialx, partialy, h
        integer n, i
        h = 10E-7
        n = (x - x0)/stepsize
        do 20 i = 0, n
                y1 = y0 + stepsize*f(x0,y0) + (stepsize**2.0)*(partialx(x0,y0,h) + f(x0,y0)*partialy(x0,y0,h))
                x1 = x0 + stepsize
                x0 = x1
                y0 = y1
        20 continue
        t = y1
end function taylor_method

function adam_bashford_method(x0,y0,x,stepsize) result(a)
        double precision f, x0, y0, x, stepsize, x1, y1, x2, y2, a
        integer n, i
        n = (x - x0)/stepsize
        x1 = x0 + stepsize
        y1 = y0 + stepsize*f(x0,y0)
        !print*,x1
        !print*,y1
        do 30 i = 1 ,n
                y2 = y1 + stepsize*(1.5*f(x1,y1) - 0.5*f(x0,y0))
                x2 = x1 + stepsize
                x0 = x1
                x1 = x2
                y0 = y1
                y1 = y2
        30 continue
        a = y2
end function adam_bashford_method

function runge_kutta(x0,y0,x,stepsize) result(r)
        double precision f, x0, y0, x, x1, y1, stepsize, k, r
        integer n, i
        n = (x - x0)/stepsize
        do 40 i = 1,n
                k = stepsize*f(x0,y0)
                y1 = y0 + stepsize*f(x0 + stepsize*0.5, y0 + k*0.5)
                x1 = x0 + stepsize
                x0 = x1
                y0 = y1
        40 continue
        r = y1
end function runge_kutta

program solve_diff_eqn

implicit none
double precision x0,y0,x,stepsize, y_real_1, e, t, a, r, y_real_3
double precision euler_method, taylor_method, adam_bashford_method, runge_kutta 

y_real_1 = EXP(-0.5)
y_real_3 = EXP(-4.5)

x0 = 0
y0 = 1
x = 3
stepsize = 0.001

e = euler_method(x0, y0, x, stepsize)

print*, e
print*, e - y_real_3
print*,"###########################################"

x0 = 0
y0 = 1
t = taylor_method(x0, y0, x, stepsize)

print*,t
print*,t - y_real_3
print*,"###########################################"

x0 = 0
y0 = 1
a = adam_bashford_method(x0, y0, x, stepsize)

print*,a
print*,a - y_real_3
print*,"###########################################"

x0 = 0
y0 = 1
r = runge_kutta(x0, y0, x, stepsize)

print*,r
print*,r - y_real_3

end program solve_diff_eqn

File: Homework_3/diff_eqn_solver.py
import numpy as np
from astropy.table import Table
from astropy.io import ascii

## Comparing different algorithms to solve dy/dx = f(x,y)

def euler_method(function,x0,y0,x,stepsize):
    """
    function = dy/dx; x0,y0 are initial conditions; x is the point where y needs to be calculated.
    """
    x_array = np.arange(x0, x + stepsize, stepsize)
    y_array = np.zeros(len(x_array))
    y_array[0] = y0
    for i in range(1,len(x_array)):
        y_array[i] = y_array[i-1] + stepsize*function(x_array[i-1],y_array[i-1])
    
    return y_array[-1]

def partialx(function,x0,y0,stepsize):
    """
    Returns partial first derivative of the function "function" at the point x0,y0 with respect to x by considering points, one and two steps on either side of x0.
    The Accuracy is of order (stepsize)^4.
    """
    return (function(x0 - 2*stepsize, y0) - 8*function(x0 - stepsize, y0) + 8*function(x0 + stepsize, y0) - function(x0 + 2*stepsize, y0))/(12*stepsize)

def partialy(function,x0,y0,stepsize):
    """
    Returns partial first derivative of the function "function" at the point x0,y0 with respect to y by considering points, one and two steps on either side of y0.
    The Accuracy is of order (stepsize)^4.
    """
    return (function(x0, y0 - 2*stepsize) - 8*function(x0, y0 - stepsize) + 8*function(x0, y0 + stepsize) - function(x0, y0 + 2*stepsize))/(12*stepsize)


def taylor_method(function,x0,y0,x,stepsize):
    """
    function = dy/dx; x0,y0 are initial conditions; x is the point where y needs to be calculated.
    """
    x_array = np.arange(x0, x + stepsize, stepsize)
    y_array = np.zeros(len(x_array))
    y_array[0] = y0
    for i in range(1,len(x_array)):
        y_array[i] = y_array[i-1] + stepsize*function(x_array[i-1],y_array[i-1]) + (stepsize**2.0)*(partialx(function,x_array[i-1],y_array[i-1],10**(-7)) + function(x_array[i-1],y_array[i-1])*partialy(function,x_array[i-1],y_array[i-1],10**(-7)))
    
    return y_array[-1]

def adam_bashford_method(function,x0,y0,x,stepsize):
    """
    function = dy/dx; x0,y0 are initial conditions; x is the point where y needs to be calculated.
    """
    x_array = np.arange(x0, x + stepsize, stepsize)
    y_array = np.zeros(len(x_array))
    y_array[0] = y0
    y_array[1] = y0
    for i in range(2,len(x_array)):
        y_array[i] = y_array[i-1] + stepsize*(1.5*function(x_array[i-1],y_array[i-1]) - 0.5*function(x_array[i-2],y_array[i-2]))

    return y_array[-1]

def runge_kutta_method(function,x0,y0,x,stepsize):
    """
    function = dy/dx; x0,y0 are initial conditions; x is the point where y needs to be calculated.
    """
    x_array = np.arange(x0, x + stepsize, stepsize)
    y_array = np.zeros(len(x_array))
    y_array[0] = y0
    for i in range(1,len(x_array)):
        y_step = stepsize*function(x_array[i-1], y_array[i-1])
        y_array[i] = y_array[i-1] + stepsize*function(x_array[i-1] + (stepsize/2.0), y_array[i-1] + (y_step/2.0))
    
    return y_array[-1] 

def f(x,y):
    return -x*y

x0 = 0.0
y0 = 1.0
x1 = 1.0
x3 = 3.0

x = np.array([1.0,3.0])

print(euler_method(f,x0,y0,np.all(x),0.001))

stepsize_array = np.array([0.5, 0.2, 0.1, 0.05, 0.02, 0.01, 0.005, 0.002, 0.001])

#e1_array = np.zeros(len(stepsize_array))
#t1_array = np.zeros(len(stepsize_array))
#ab1_array = np.zeros(len(stepsize_array))
#rk1_array = np.zeros(len(stepsize_array))
#
#e3_array = np.zeros(len(stepsize_array))
#t3_array = np.zeros(len(stepsize_array))
#ab3_array = np.zeros(len(stepsize_array))
#rk3_array = np.zeros(len(stepsize_array))
#
#e1_diff = np.zeros(len(stepsize_array))
#t1_diff = np.zeros(len(stepsize_array))
#ab1_diff = np.zeros(len(stepsize_array))
#rk1_diff = np.zeros(len(stepsize_array))
#
#e3_diff = np.zeros(len(stepsize_array))
#t3_diff = np.zeros(len(stepsize_array))
#ab3_diff = np.zeros(len(stepsize_array))
#rk3_diff = np.zeros(len(stepsize_array))
#
#real_value_1 = np.exp(-1.0/2.0)
#real_value_3 = np.exp(-9.0/2.0)
#
#for i in range(len(stepsize_array)):
#    e1_array[i] = "{a:0.6f}".format(a=euler_method(f,x0,y0,x1,stepsize_array[i]))
#    e3_array[i] = "{a:0.6f}".format(a=euler_method(f,x0,y0,x3,stepsize_array[i]))
#    e1_diff[i] = "{b:0.3E}".format(b=abs(euler_method(f,x0,y0,x1,stepsize_array[i]) - real_value_1))
#    e3_diff[i] = "{b:0.3E}".format(b=abs(euler_method(f,x0,y0,x3,stepsize_array[i]) - real_value_3))
#
#    t1_array[i] = "{a:0.6f}".format(a=taylor_method(f,x0,y0,x1,stepsize_array[i]))
#    t3_array[i] = "{a:0.6f}".format(a=taylor_method(f,x0,y0,x3,stepsize_array[i]))
#    t1_diff[i] = "{b:0.3E}".format(b=abs(taylor_method(f,x0,y0,x1,stepsize_array[i]) - real_value_1))
#    t3_diff[i] = "{b:0.3E}".format(b=abs(taylor_method(f,x0,y0,x3,stepsize_array[i]) - real_value_3))
#
#    ab1_array[i] = "{a:0.6f}".format(a=adam_bashford_method(f,x0,y0,x1,stepsize_array[i]))
#    ab3_array[i] = "{a:0.6f}".format(a=adam_bashford_method(f,x0,y0,x3,stepsize_array[i]))
#    ab1_diff[i] = "{b:0.3E}".format(b=abs(adam_bashford_method(f,x0,y0,x1,stepsize_array[i]) - real_value_1))
#    ab3_diff[i] = "{b:0.3E}".format(b=abs(adam_bashford_method(f,x0,y0,x3,stepsize_array[i]) - real_value_3))
#
#    rk1_array[i] = "{a:0.6f}".format(a=runge_kutta_method(f,x0,y0,x1,stepsize_array[i]))
#    rk3_array[i] = "{a:0.6f}".format(a=runge_kutta_method(f,x0,y0,x3,stepsize_array[i]))
#    rk1_diff[i] = "{b:0.3E}".format(b=abs(runge_kutta_method(f,x0,y0,x1,stepsize_array[i]) - real_value_1))
#    rk3_diff[i] = "{b:0.3E}".format(b=abs(runge_kutta_method(f,x0,y0,x3,stepsize_array[i]) - real_value_3))
#
#
#t1 = Table([stepsize_array, e1_array, e1_diff, t1_array, t1_diff, ab1_array, ab1_diff, rk1_array, rk1_diff], names=["h", "e1", "e1_diff", "t1", "t1_diff", "ab1", "ab1_diff", "rk1", "rk1_diff"])
#t3 = Table([stepsize_array, e3_array, e3_diff, t3_array, t3_diff, ab3_array, ab3_diff, rk3_array, rk3_diff], names=["h", "e3", "e3_diff", "t3", "t3_diff", "ab3", "ab3_diff", "rk3", "rk3_diff"])
#
#ascii.write(t1,"final_table_x1.tex",format="latex")
#ascii.write(t3,"final_table_x3.tex", format="latex")

#print("Real value : ",np.exp(-1.0/2.0))
#print("Euler method : ",euler_method(f,0.0,1.0,1.0,0.001))
#print("Taylor method : ",taylor_method(f,0.0,1.0,1.0,0.001))
#print("Adam Bashford method : ",adam_bashford_method(f,0.0,1.0,1.0,0.001))
#print("Runge Kutta method : ",runge_kutta_method(f,0.0,1.0,1.0,0.001))
